<a href="https://colab.research.google.com/github/Gise05624/Proyecto_DataScience_I/blob/main/Entrega_3/ProyectoParteIII_Bargas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Proyecto Final

### Archivo: ProyectoDS_ParteI_Bargas.ipynb
_______________________________________

# ✔ Modelado y Validación

Este proyecto forma parte del análisis comercial de un conjunto de datos con información mensual sobre ventas, costos operativos y satisfacción del cliente, segmentado por región y tipo de producto. En esta tercera parte busco aplicar técnicas de selección de variables y modelado predictivo para estimar el volumen de ventas.

_______________________________________

## 🗂️📈 Preparación de los datos

Se cargan los datos, se codifican las variables categóricas y se convierten las fechas en variables numéricas para su uso en modelos.



In [2]:

import pandas as pd

df = pd.read_csv('ventas_dataset_actdeclase.csv')
df['Mes'] = pd.to_datetime(df['Mes'])
df_modelo = df.copy()
df_modelo = pd.get_dummies(df_modelo, columns=['Region', 'Producto'], drop_first=True)
df_modelo['Mes'] = df_modelo['Mes'].dt.month
df_modelo.head()


Mes  Ventas  Satisfaccion_Cliente  Costo_Operativo  Region_Norte  \
0    1    4174              4.593086             1794          True   
1    2    2130              4.559382             2138          True   
2    3    3169              3.311989              830          True   
3    4    2482              4.732352              630          True   
4    5    2685              3.112823             2891          True   

   Region_Oeste  Region_Sur  Producto_Producto_B  Producto_Producto_C  
0         False       False                False                False  
1         False       False                False                False  
2         False       False                False                False  
3         False       False                False                False  
4         False       False                False                False

____________________

## 📊 Selección de variables (Feature Selection)

Se usa el método **SelectKBest** con función `f_regression` para seleccionar las 5 variables más relevantes para predecir las ventas.

In [3]:

from sklearn.feature_selection import SelectKBest, f_regression
from sklearn.model_selection import train_test_split

X = df_modelo.drop('Ventas', axis=1)
y = df_modelo['Ventas']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

selector = SelectKBest(score_func=f_regression, k=5)
X_train_selected = selector.fit_transform(X_train, y_train)
X_test_selected = selector.transform(X_test)

selected_features = X_train.columns[selector.get_support()].tolist()
print("Variables seleccionadas:", selected_features)


Variables seleccionadas: ['Satisfaccion_Cliente', 'Costo_Operativo', 'Region_Norte', 'Region_Oeste', 'Producto_Producto_C']


__________________________


## 👩🏻‍💻 Entrenamiento del modelo

Se utiliza un modelo de regresión basado en Random Forest para predecir el volumen de ventas.


In [4]:

from sklearn.ensemble import RandomForestRegressor

model = RandomForestRegressor(random_state=42)
model.fit(X_train_selected, y_train)

y_pred = model.predict(X_test_selected)


____________________

## 🧐 Evaluación del modelo

Se evalúa el modelo utilizando métricas de regresión: **MAE**, **RMSE** y **R²**.


In [6]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

mae = mean_absolute_error(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)
rmse = mse ** 0.5
r2 = r2_score(y_test, y_pred)

print("MAE:", mae)
print("RMSE:", rmse)
print("R²:", r2)

MAE: 1048.7724137931034
RMSE: 1177.6051911664065
R²: -0.15432475616027408


___________________

## 🔍 Conclusiones

Después de entrenar el modelo y evaluar sus resultados, se puede decir lo siguiente:

- Las variables que más ayudaron a explicar las ventas fueron: Satisfaccion_Cliente, Costo_Operativo, Region_Norte, Region_Oeste, Producto_Producto_C.
- El error promedio que tuvo el modelo (MAE) fue de aproximadamente **1048.77**, y el RMSE fue de **1177.61**.
- El valor R² fue de **-0.15**, lo cual indica que el modelo no logró explicar muy bien las variaciones en las ventas.

Aunque los resultados no fueron los mejores, este ejercicio fue útil para identificar qué variables internas parecen tener mayor impacto en las ventas. También sirve como punto de partida para seguir ajustando el modelo o sumar nuevas variables (como promociones, competencia, estacionalidad)
